storage/src/data_access_layer/balance.rs

```diff
@@ -307,13 +407,177 @@ impl BalanceRequest {
     }
 }
 
+pub trait AvailableBalanceChecker {
+    /// Calculate and return available balance.
+    #[allow(clippy::result_unit_err)]
+    fn available_balance(
+        &self,
+        block_time: BlockTime,
+        total_balance: U512,
+        gas_hold_balance_handling: GasHoldBalanceHandling,
```

Comment:
This being a trait I don't think it's very clean to pass the concrete balance handling variants here since it's not very extensible in the future. Not blocking the PR on this but we may consider refactoring this in the future (and maybe marking this as `pub(crate)`.